### DISCUSSION TITLE
Doxygen actions and in-function-body lambdas

### DISCUSSION THREAD

**Discussion by 10tailesSeer** [doxygen/doxygen#11098] [Q&A]
I don't want to rush with creating issues, since it'll be my first time, and would like to discuss it thoroughly firstly.  
So far, I can separate problem into two topics:
- `doxygen` doesn't recognize anything as a symbol (be it method, class, variable etc.) inside the function body OR I can't understand how to make it to...
- my attempts to workaround it fail when I try to combine them with automated documentation and `github actions` (to deploy gh-pages and so on)  

## Why this is important
As time goes C++ lambdas become more and more used: for reasons of DRY idiom, cleaning up class interfaces and just whatnot.
So It'd be ducking wonderful, an event of the millenia, if we were able to automatically recognize them, and properly document them. As there won't be helper functions in classes anymore, yet these lambdas should be documented somehow as well.  

### Recognizing at least something inside function body  
Is there a command I haven't tried yet, or used correctly? So far I tried to force appearing anything new with `\fn`, `\property`, post-documentation. But it just ruins everything: function within which body I try to make such a declaration quits to be recongnized as one along with any other comments there.  

But I thought why just not use `\section`, `\snippet` and other goodies. And here's another problem...  

### Workarounding when things are automated  
With `actions` (including `doxygen action`), there're created a bunch of several `.html` files deployed to a dedicated branch. And when `source_browsing` is turned on, the references from recognized symbols are automatically created to generate `<modified_name>_source.html` files which is super great! But I can't find a way to create an anchor (niether by the built-in means, e.g. `\link` and `\anchor`, nor by html commands, e.g. `<a href=" ">` and `<a name=" ">` to which I could refer from freshly created sections in the documentation of the given function. The reason seems to be that when `<name>_source.html` files are created, anything within original source code file (that doxygen hasn't recognized) is treated as plain texts or even stripped off (if `strip_off_comments` is turned on).  

So the only solution i got so far is to directly create links using line number in the final `<name>_source.html` file. Which is terrible. I hope it's obvious why.

**Comment by albert-github** [comment]
Doxygen is documenting the outside of the functions etc. which should describe the functioning of the of a function to a user. A lambda is something that is inside one function and in principle of no interest to the end user, though for developers it can be useful also to document the inside of a function. For the later the setting `INTERNAL_DOCS` and the `\internal` command exists. So please have a look at these.

Regarding the "But I can't find a way to create an anchor..." this is impossible yo tell what the problem here might be.

Like written with the comment you initially posted:

- Can you please attach a, small, self contained example (source+configuration file in a, compressed, tar or zip file!) that allows us to reproduce the problem? Please don't add external links as they might not be persistent (also references to GitHub repositories are considered non persistent).
- Please also specify the full doxygen version used (doxygen -v).

**Comment by 10tailesSeer** [comment]
Tried `\internal`, didn't help with links. But `hide_in_body_docs = YES` seemed to do equivalent  job extracting-docs-wise.

Doxygen version: 1.12.0 (the desription of `edge tag` in doxygen action docs are invalid)  
[doxygen_issue.zip](https://github.com/user-attachments/files/16740412/doxygen_issue.zip)

Remember that from `gh-pages` branch one has only access to produced html files.
I also thought of using `\lineinfo` but couldn't come up with an idea how to shove that in a html-`href` commant: every way I tried so far to directly or indirectly insert its value into `<a href="">` instantly triggers smth and `\lineinfo` isn't recognized as a doxy command anymore

**Reply by albert-github** [reply]
The Doxyfile in the zip file says it has been generated with doxygen version 1.12.0 (so the current doxygen version).
You can always get the version of doxygen by means of `doxygen -v` and also the version of doxygen is normally in a meta tag of the generated html files (and on a few other places in the files).

The main problem that occurs is that the anchor definitions have to be in doxygen comment sections (either in the form of `<a name=...` ore `\anchor` or ... and that this results in a link to the documentation and not tho the source code.
It is correct that `<a href="mre_8cpp_source.html#l00015">` will work but the remark: "but it's absolute, thus any changes in the source will mess it up" is 100% correct.

It looks like a new command is required here `\source_anchor` (just a name):
```
\source_anchor["{"<explaining text>"}"]  <anchor>
```
this could be reference as:
```
\ref <anchor>
```
giving an link to the source code line and as line text the text from `<explaining text>`
or alternativemy
```
\ref <anchor> "new text"
```
giving an link to the source code line and as line text the text specified `new text`
Of cource the source code browser for the specific file has to be enabled (and otherwise doxygen should do this under the hood.

@doxygen what do you think or do you have an other idea?

**Reply by 10tailesSeer** [reply]
About the version: yeaah, i noticed now, it's really 1.12.0. Well, then [description](https://github.com/marketplace/actions/doxygen-action#supported-tags) is actually wrong.

About another failed workaround: I tried to shove `\lineinfo` with custom commands, in the form:
```
ALIASES += l{2}="\1\2"
*somewhere in the midst of codelines*
///<a href= \l{mre_8cpp_source.html#l000,\lineinfo} >linktext</a>
```  
but the exact moment `<`  is encountered, `\lineinfo` stops being parsed as doxy command. And nothing changes whether I try to escape it, remove to alias definition, add another level of indirection or whatnot

**Reply by albert-github** [reply]
> About the version: yeaah, i noticed now, it's really 1.12.0. Well, then [description](https://github.com/marketplace/actions/doxygen-action#supported-tags) is actually wrong.

I see that in their repository (not maintained by doxygen) https://github.com/mattnotmitt/doxygen-action  there is already an issue for the versioning.

Regarding the
```
ALIASES += l{3}="\1\2\3"
*somewhere in the midst of codelines*
/// \l{<a href=\"<filename>_source.html#,\lineinfo,\">link text</a>}
```


- `<fileinfo>` syntax is not correct here should be something like `\fileinfo{name}`
- `\lineinfo` this will just give a number (like `15`) but not something like `l00015`
- all would still fail as the how would you reference this in the documentation the `\fileinfo` / `\lineinfo` return information about the place where they are issued.

**Reply by 10tailesSeer** [reply]
I'm sorry, copied commands badly, updated now. 
The things you mentioned: I was about smth else. Everything works quite well actually: link is generated correctly when without `<`-symbol. But as `<` appears `\lineinfo isn't parsed correctly. And it's the same problem when using markdown format. Why???

**Reply by albert-github** [reply]
> And it's the same problem when using markdown format. Why???

The `\fileinfo` will refer to the current file and thus not to the source but the documentation.
